Extend service usage APIs with costs and free allowance

benthorner opened the pull request:
https://www.pivotaltracker.com/story/show/181934027

Admin uses a couple of APIs to get data to show yearly
and monthly usage info on dashboard pages:

  - /service/{0}/billing/monthly-usage
  - /service/{0}/billing/yearly-usage-summary

Both of these APIs return a bunch of JSON rows, grouped
by "rate", "notification_type" (and "month" for monthly
usage). The /monthly-usage API also groups by "postage"
so we can show a detailed breakdown for each month [^1].

This PR adds a bunch of new fields to replace some of the 
manual calculations in the Admin app, rather than trying to 
give them more data to cope with variable SMS rates.

- /yearly-usage-summary:
  - chargeable_units (to replace billing_units)
  - cost
  - free_chargeable_units

- /monthly-usage:
  - chargeable_units (to replace billing_units)
  - cost
  - free_chargeable_units

Each API picks the data it needs from a single query for 
each notification type, ensuring the calculations for the 
common fields are consistent in future.

Please review commit by commit. The first commits are 
a bunch of refactorings so the main changes are smaller 
and simpler.

## Testing

Given the complexity of billing I also did some manual testing 
with two services:

- [SMS example](https://www.notifications.service.gov.uk/services/e5bb3a40-7a7a-467c-b937-634ae5b19539/usage?year=2022)
- [Letter example](https://www.notifications.service.gov.uk/services/c003bfd7-8079-4c0d-9883-37d92727832c/usage?year=2022)
- [Additional SMS example](https://www.notifications.service.gov.uk/services/c5956607-20b1-48b4-8983-85d11404e61f/usage?year=2021)

To do the tests locally I copied anonymised billing data for them 
into my local DB e.g.

```
# remotely
\copy (select * from annual_billing where service_id in ('e5bb3a40-7a7a-467c-b937-634ae5b19539', 'c003bfd7-8079-4c0d-9883-37d92727832c', 'c5956607-20b1-48b4-8983-85d11404e61f')) to 'annual_billing.csv' delimiter ',' csv header;

# locally
alter table annual_billing drop constraint annual_billing_service_id_fkey;
 
copy annual_billing(id,service_id,financial_year_start,free_sms_fragment_limit,updated_at,created_at) from '/Users/ben.thorner/notify/credentials/annual_billing.csv' delimiter ',' csv header;
```

For convenience I also removed [the auth restrictions](https://github.com/alphagov/notifications-api/blob/b4ffcac3539cb1dabbb20e09af5c5722a8d8612a/app/__init__.py#L250) and [the code
that auto updates ft_billing](https://github.com/alphagov/notifications-api/blob/b4ffcac3539cb1dabbb20e09af5c5722a8d8612a/app/dao/fact_billing_dao.py#L258) (depends on the former).

Then I could curl the APIs locally and compare before / after e.g.

```
curl "http://localhost:6011/service/e5bb3a40-7a7a-467c-b937-634ae5b19539/billing/monthly-usage?year=2020"
```

Things I checked (for both APIs):

- The number of rows is the same before and after.
- "billing_units" are the same before and after.
- "cost" matches what is shown in Admin app now*.
- Partial costs appear as expected (free + paid).

\* For letters this involves manually adding up across all postage rates, since we have a row for each rate.

[^1]: https://www.notifications.service.gov.uk/services/c003bfd7-8079-4c0d-9883-37d92727832c/usage

idavidmcdonald commented on tests/app/dao/test_ft_billing_dao.py:
```diff
@@ -531,24 +552,30 @@ def test_fetch_billing_totals_for_year_variable_rates(notify_db_session):
     assert results[3].chargeable_units == 5
     assert results[3].rate == Decimal('0.162')
+    # 1 free unit on the 16th, 1 on the 17th (+ 3 paid)
+    assert results[3].free_allowance_remainder == 0
```
I'm a bit confused what the `free_allowance_remainder` is

the `notifications_sent`, `billable_units` and `chargeable_units` are all things that are counting to do with the `rate`. ie how many things you've sent at a certain rate. But it doesn't make intuitive sense to me to have a `free_allowance_remainder` attributed to a rate. I can see it only being attributed to a point in time, that is unless you are using the rate change as proxy for the date? If so, this seems like a particularly complex thing to get your head around?

This is just a bit of a quick comment before I rush off to meetings but hoping to get you thinking about it/replying in time for all my meetings to end if I'm lucky ;)

benthorner replied:
(For anyone else reviewing, don't worry about this as it's gone.)

As discussed, I've removed this field / commit but added "free_chargeable_units" to the yearly usage API in the final commit, as we can still benefit from an explicitly named field to subtract with.

leohemsted reviewed (commented):
I've got a couple of reasonably large thoughts/concerns about this, but I also accept the time constraints and don't want us to end up with a fundamentally broken usage page at the beginning of next week. Keen to sit down and work through how best to solve this, both immediately and then in the medium term.

benthorner commented on app/dao/fact_billing_dao.py:
```diff
@@ -197,111 +197,164 @@ def fetch_letter_line_items_for_all_services(start_date, end_date):
+        func.sum(chargeable_units).over(
+            order_by=[
+                FactBilling.bst_date,  # order is "ASC" by default
+                FactBilling.rate  # ensures test stability for rows on the same day
```
It turns out we don't need to worry about `rate` or any of the other fields when we're aggregating on a daily basis: as long as the rate is unchanged the order of the rows doesn't matter, because the rates are all the same.

I've amended the "cost" commit to make this assumption clear and removed the redundant order column

benthorner commented on app/dao/fact_billing_dao.py:
```diff
@@ -197,111 +197,164 @@ def fetch_letter_line_items_for_all_services(start_date, end_date):
+                FactBilling.bst_date,  # order is "ASC" by default
+                FactBilling.rate  # ensures test stability for rows on the same day
+            ],
+            rows=(None, -1)  # ROWS BETWEEN UNBOUNDED PRECEDING AND 1 ROW PRECEDING
```
As discussed, I've revised this function with a bunch more documentation - see the "cost" commit.

benthorner commented on app/dao/fact_billing_dao.py:
```diff
@@ -197,111 +197,164 @@ def fetch_letter_line_items_for_all_services(start_date, end_date):
+        func.least(
+            cumulative_free_remainder,
+            chargeable_units,
+        ).cast(Integer).label("free_chargeable_units"),  # for some reason the result is a Decimal
```
As discussed, I've revised the names. In particular, `free_chargeable_units` is now `free_allowance_used` (I thought "used" was slightly more friendly than "consumed").

Just a note for anyone else reading: this is complex no matter how we implement it, but the situation was made worse by the unclear variable names and limited documentation, which I've addressed.

benthorner commented on app/billing/billing_schemas.py:
```diff
@@ -12,30 +12,36 @@
+            "rate": float(row.rate),
+            "postage": row.postage,
+            "cost": float(row.cost),
+            "free_chargeable_units": row.free_chargeable_units,
```
I'm going to resolve this one for now based on other changes we discussed.

The situation is indeed very confusing - we do something different for each notification type!

benthorner commented:
@leohemsted thanks for the chat earlier and in general for the effort to review this!

I've fixed up (so many rebases!) the changes we discussed earlier, as well as various additional fixes after trying to use this in Admin. I've also re-tested the API output using the three examples in the description.

The first few commits are untouched. Main changes:

- Some new commits to fix some misleading test data, before the API changes.
- I've combined the yearly + monthly changes into single commits.
- Added a load of documentation and tweaked names here and there.
- Changed the "outerjoin" to a normal "join".

When you have some time to take another look, I'd recommend reviewing commit by commit from "Fix incorrect billable_units for email test data" onwards.

benthorner commented:
Nice, thanks @leohemsted !

I've already got a follow-up note to simplify some of the tests, so hopefully that'll help there.

Regarding the naming, I had a go at changing the names but couldn't convince myself they were significantly better. I find "this row" more confusing at the point of definition, where it isn't in the context of a row. Happy to look at a follow-up PR if we want to try and unify the terminology later on though.
